spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/dsl/Stream.java

```diff
@@ -0,0 +1,250 @@
+package org.springframework.cloud.dataflow.rest.client.dsl;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import org.springframework.cloud.dataflow.rest.client.DataFlowOperations;
+import org.springframework.cloud.dataflow.rest.resource.StreamDefinitionResource;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+
+/**
+ * @author Vinicius Carvalho
+ */
+public class Stream {
+
+	Stream(String name, List<StreamApplication> applications, DataFlowOperations client) {
+		this.name = name;
+		this.applications = applications;
+		this.client = client;
+	}
+
+	Stream(String name, String definition, DataFlowOperations client){
+		this.name = name;
+		this.definition = definition;
+		this.client = client;
+	}
+
+	private String name;
+
+	private List<StreamApplication> applications = new LinkedList<>();
+
+	private String definition;
+
+	private DataFlowOperations client;
+
+	private String status;
+
+	/**
+	 * Resolves the name of a named channel for an application. For example, if a stream called foo has an application time with a label tick
+	 * the result would be :foo.tick
+	 *
+	 * @param applicationName The application name to resolve the named destination
+	 * @return the named channel to use to tap on an application
+	 */
+	public String namedChannelFor(String applicationName){
+		for(StreamApplication app : applications){
+			if(app.getName().equals(applicationName)){
+				String id =  StringUtils.isEmpty(app.getLabel()) ? app.getName() : app.getLabel();
+				return String.format(":%s.%s", this.name, id);
+			}
+		}
+		return null;
+	}
+
+	public String getDefinition(){
+		if(StringUtils.isEmpty(this.definition)){
+			createStreamDefinition();
+		}
+		return this.definition;
+	}
+
+	public Stream create() {
+		StreamDefinitionResource definition = this.client.streamOperations().createStream(this.name, getDefinition(), false);
+		this.status = definition.getStatus();
+		return this;
+	}
+
+	public Stream deploy(Map<String, String> deploymentProperties){
+		Map<String, String> resolvedProperties = resolveDeploymentProperties(deploymentProperties);
+		if(this.status == null){
```

Comment:
It might be better not to keep any state in this class as it will always be out of sync even if you call `refreshStatus`. For example, create to exactly same dsl's using different builder instances, one of those is always out of sync. You'll end up chasing your own tail.

Reply:
Yes keeping in sync will be a challenge but the state is there for a few reasons:

1. We need to know whether we are deploying using a definition bypassing the fluent api, that was a request from Mark
2. In order to support a deploy/create style, for instance to call `create(name, definition, true)` we need to see if there is already a status from a create call, or not.

Reply:
In the sample above
```
Stream ticktock = Stream.Builder(template)
    .name("ticktock")
    .from(time)
    .to(new StreamApplication("log")
    .build();
```
It looks silly to do `new StreamApplication("log")` because no other properties on the StreamApplication were set.  This example would be better represented using the `definition` method on the build

```
		Stream stream = Stream.Builder(client)
				.name("ticktock")
				.definition("time | log")
				.deploy(Collections.singletonMap("deployer.log.count","2"));

```
The idea suggested by @mminella was that we have have the apps declared as `@Bean` in an `@Configuration` class and then the DSL is mostly concerned about wiring them up together and providing deployment properties.  This makes it much easier to reuse the StreamApplication instances across different stream definitions and to more easily see what the stream is doing vs. the DSL style where each individual app was created+properties set+stitched together with from/via/to.

Reply:
I am not remembering the discussion regarding storing the status.   
```
We need to know whether we are deploying using a definition bypassing the fluent api, that was a request from Mark
```
If I was in the shell and deployed 'ticktock' and then executed an application using the DSL which also tried to create `ticktock`, I should get `DuplicateStreamDefinitionException` thrown back to the client.  Can you remind me what I was saying.... ;)  I can certainly see the argument for not storing the state.

```
In order to support a deploy/create style, for instance to call create(name, definition, true) we need to see if there is already a status from a create call, or not.
```
It would be possible to work around this by requiring the users to write

```
		Stream stream = Stream.Builder(client)
				.name("ticktock")
				.definition("time | log")
                                .create()
				.deploy(Collections.singletonMap("deployer.log.count","2"));
```
That is, require the `.create()` method to be called.  This whole shortcut of `--deploy` on create has kind of been a PITA when it came to Skipper integration.  We used to have `stream create BLAH --deploy --useSkipper` as a shortcut but it started to get complex/confusing.  So now, we decided to have a new command `stream skipper deploy` to cleanly separate out the skipper variant of deploying.  To deploy using skipper, you always have to create the stream and then deploy it in two steps.   Following the same approach here would seem to work well, contact is 'clean' and would avoid the need for state to be stored.

Which reminds me... we will have to adopt the DSL to take into account skipper...